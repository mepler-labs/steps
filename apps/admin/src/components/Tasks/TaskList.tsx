import React from 'react';
import { Flex, Box } from 'grid-styled';
import {SortableContainer, SortableElement, arrayMove} from 'react-sortable-hoc';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { getTasks, setTasks } from 'actions/tasks';
import styled from 'styled-components';
import Task from './TaskListItem';
import NoTasks from './NoTasks';

interface Props {
  className?: string;
  actions: any;
  tasks: any;
}

const TaskContainer = styled.div`
  border: 1px solid blue;
  margin: 1em 0;
  padding-left: 10%;
  position: relative;

  .number {
    border-right: 1px solid blue;
    bottom: 0;
    font-size: 2em;
    left: 0;
    padding: .5em 0;
    position: absolute;
    text-align: center;
    top: 0;
    width: 10%;
  }
`;

const SortableList = SortableContainer(({ items }) => {
  return (
    <Box>
      {items.map((task, index) => (
        <TaskContainer>
          <div className='number'>{index + 1}</div>
          <Task key={`item-${index}`} index={index} value={task.content} />
        </TaskContainer>
      ))}
    </Box>
  );
});

class TaskList extends React.Component<Props, {}> {
  componentWillMount () {
    this.props.actions.getTasks();
  }

  onSortEnd = ({oldIndex, newIndex}) => {
    this.props.actions.setTasks(
      arrayMove(this.props.tasks, oldIndex, newIndex),
    );
  };

  render () {
    const tasks = this.props.tasks;
    const taskDisplay = tasks.length > 0 ? (
        <Box width={1} p={4}>
          <h2>Tasks</h2>
          <Link to="/clients/1/tasks/new">New Task</Link>
          <SortableList items={this.props.tasks} onSortEnd={this.onSortEnd} />
        </Box>
    ) : (
        <NoTasks></NoTasks>
      );

    return (
      <div>
        {taskDisplay}
      </div>
    );
  }
}

const StyledTaskList = styled(TaskList)`
  display: flex;
`;

const mapStateToProps = (state, props) => ({
  tasks: state.tasks.tasks,
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators({ getTasks, setTasks }, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(StyledTaskList);
