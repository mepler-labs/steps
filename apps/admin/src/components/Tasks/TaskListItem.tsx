import { Box, Flex } from 'grid-styled';
import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import { blue, green, lightGrey, white } from 'styles/colors';
import { remCalc, sansSerif } from 'styles/type';

const StyledLink = styled(Link)`
  color: ${green};
  font-size: 0.8em;
  margin-top: 0.5em;
  text-decoration: none;
  text-transform: uppercase;
`;

export default ({ setTaskStatus, task, url, index }) => {
  const toggleTaskStatus = e => {
    const status = task.status === 'COMPLETED' ? 'ACTIVE' : 'COMPLETED';
    setTaskStatus(task, status);
  };

  const checked =
    task.status === 'COMPLETED' ? 'check_circle' : 'check_circle_outline';

  return (
    <Background
      flexDirection="column"
      alignItems="flex-start"
      justifyContent="center"
      p={20}
      width={1}
    >
      <Flex flexDirection="row" alignItems="center">
        <div onClick={toggleTaskStatus}>
          <i className={`material-icons ${checked}`}>{checked}</i>
        </div>
        <Title>{task.title}</Title>
      </Flex>
      <Box ml={34}>
        <StyledLink to={`${url}/${task.id}`}>View Steps &rarr;</StyledLink>
      </Box>
    </Background>
  );
};

const Background = styled(Flex)`
  background: ${white};
  border-bottom-right-radius: 4px;
  border-top-right-radius: 4px;

  i.material-icons {
    cursor: pointer;
    width: 24px;
  }

  .check_circle {
    color: ${blue};
  }

  .check_circle_outline {
    color: ${lightGrey};
  }
`;

const Title = styled.h3`
  font-family: ${sansSerif};
  font-size: ${remCalc(24)};
  font-weight: 500;
  margin-top: 0;
  margin-left: 10px;
  margin-right: 0;
  margin-bottom: ${remCalc(10)};
`;
