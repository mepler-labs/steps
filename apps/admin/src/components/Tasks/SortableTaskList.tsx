import React from 'react';
import { connect } from 'react-redux';
import { Link, Match } from 'react-router-dom';
import { arrayMove, SortableContainer } from 'react-sortable-hoc';
import { bindActionCreators } from 'redux';
import { Box, Flex } from 'grid-styled';
import map from 'lodash/map';
import uniq from 'lodash/uniq';

import { filterById, findById } from 'helpers';
import Button from 'atoms/Buttons/Button';
import Modal from 'containers/Modal';
import TermsModal, { TERMS } from 'components/Clients/TermsModal';
import { ModalSize } from '../Modal';
import Filter, { FilterCategory } from 'components/Filter';
import NoTasks from './NoTasks';

import { getTasks, orderTasks, setTaskStatus } from 'actions/tasks';
import { hideModal, showModal } from 'actions/modals';
import TaskList from './TaskList';

interface ListProps {
  setTaskStatus;
  items: any;
  url: string;
}

const SortableList = SortableContainer((props: ListProps) => {
  return <TaskList {...props} sortable={true} />;
});

interface Props {
  actions?: any;
  tasks?: any;
  user: any;
  match: Match;
}

interface State {
  categories: FilterCategory[];
}

export class SortableTaskList extends React.Component<Props, State> {
  constructor(props) {
    super(props);
    this.state = {
      categories: [],
    };
  }

  componentDidUpdate(prevProps) {
    if (this.props.tasks === prevProps.tasks) return;
    const categories = uniq(map(this.props.tasks, 'category')).map(
      (name: string) => ({
        name,
        active: true,
      }),
    );
    this.setState({ categories });
  }

  onSortEnd = ({ oldIndex, newIndex }) => {
    this.props.actions.orderTasks(
      arrayMove(this.props.tasks, oldIndex, newIndex),
    );
  };

  shouldCancelStart = e => {
    if (
      e.target.tagName.toLowerCase() === 'a' ||
      e.target.tagName.toLowerCase() === 'i'
    ) {
      return true;
    }
  };

  updateCategories = category => {
    const categories = this.state.categories.map(c => {
      if (c.name !== category.name) return c;
      return { ...c, active: !c.active };
    });
    this.setState({ categories });
  };

  render() {
    const { tasks, user, match, actions } = this.props;
    const { categories } = this.state;
    const filteredTasks = tasks.filter(t =>
      categories.map(c => !!c.active && c.name).includes(t.category),
    );
    const taskDisplay =
      tasks.length > 0 ? (
        <Box>
          <h2>Tasks</h2>
          <Filter categories={categories} update={this.updateCategories} />
          <SortableList
            items={filteredTasks}
            onSortEnd={this.onSortEnd}
            shouldCancelStart={this.shouldCancelStart}
            setTaskStatus={this.props.actions.setTaskStatus}
            url={match.url}
          />
          <Flex justifyContent="center">
            <Link to={`/clients/${user.id}/tasks/add`}>
              <Button>Add New Task</Button>
            </Link>
          </Flex>
        </Box>
      ) : (
        <div>
          <NoTasks user={user} />
          <Modal size={ModalSize.Large} id={TERMS}>
            <TermsModal
              onClose={() => this.props.actions.hideModal(TERMS)}
              phoneNumber="+16467988004"
            />
          </Modal>
        </div>
      );
    return <div>{taskDisplay}</div>;
  }
}

export class ConnectedTaskList extends React.Component<Props, {}> {
  componentWillMount() {
    this.props.actions.getTasks().then(() => {
      if (this.props.tasks.length === 0) {
        this.props.actions.showModal(TERMS);
      }
    });
  }
  render() {
    return <SortableTaskList {...this.props} />;
  }
}

const mapStateToProps = (state, props) => {
  return {
    tasks: filterById(
      state.tasks.tasks,
      props.user.id || props.match.params.id,
      'user_id',
    ),
    user: findById(
      state.clients.clients,
      props.user.id || props.match.params.id,
    ),
  };
};

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(
    { getTasks, orderTasks, setTaskStatus, showModal, hideModal },
    dispatch,
  ),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ConnectedTaskList);
