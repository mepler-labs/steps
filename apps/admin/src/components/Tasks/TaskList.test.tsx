import { shallow } from 'enzyme';
import React from 'react';
import TaskList from './TaskList';
import TaskListItem from './TaskListItem';
import { SortableTaskListItem } from './SortableTaskList';

describe('TaskList.tsx', () => {
  it('renders correctly', () => {
    const tasks = [
      {
        title: 'My first task',
        description: 'Task description',
      },
      {
        title: 'My second task',
        description: 'Task description',
      },
    ];
    const wrapper = shallow(
      <TaskList setTaskStatus={() => {}} items={tasks} url="foo/bar">
        {props => <TaskListItem {...props} />}
      </TaskList>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('can accept list items', () => {
    const tasks = [
      {
        title: 'My first task',
        description: 'Task description',
      },
      {
        title: 'My second task',
        description: 'Task description',
      },
    ];
    const wrapper = shallow(
      <TaskList setTaskStatus={() => {}} items={tasks} url="foo/bar">
        {props => <SortableTaskListItem {...props} />}
      </TaskList>,
    );

    const items = wrapper.find(SortableTaskListItem);

    expect(items).toHaveLength(2);
  });
});
