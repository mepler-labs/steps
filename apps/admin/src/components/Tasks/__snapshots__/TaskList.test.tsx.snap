// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskList.tsx renders correctly 1`] = `
<Box
  theme={
    Object {
      "breakpoints": Array [
        "40em",
        "52em",
        "64em",
      ],
      "fontSizes": Array [
        12,
        14,
        16,
        20,
        24,
        32,
        48,
        64,
        96,
        128,
      ],
      "space": Array [
        0,
        4,
        8,
        16,
        32,
        64,
        128,
        256,
        512,
      ],
    }
  }
>
  <styled.div
    className="active"
    key="0"
  >
    <styled.div>
      <div>
        1
      </div>
      <styled.div
        i={1}
      />
    </styled.div>
    <Component
      index={0}
      key="item-0"
      setTaskStatus={[Function]}
      task={
        Object {
          "description": "Task description",
          "title": "My first task",
        }
      }
      url="foo/bar"
    />
  </styled.div>
  <styled.div
    className="active"
    key="1"
  >
    <styled.div>
      <div>
        2
      </div>
      <styled.div
        i={2}
      />
    </styled.div>
    <Component
      index={1}
      key="item-1"
      setTaskStatus={[Function]}
      task={
        Object {
          "description": "Task description",
          "title": "My second task",
        }
      }
      url="foo/bar"
    />
  </styled.div>
</Box>
`;
